### PR TITLE
gcc: Take fec and number of encoders into account

### PR DESCRIPTION
**Pull request by gondaruk**
Fix case, when `bitrate` goes slightly higher than `max-bitrate` which leads to `fec-percentage` being out of range.

### CODE REVIEW

**Comment by MathieuDuponchelle** [comment]
hrm, isn't that bitrate clamped properly at this point? If not, shouldn't we fix that instead?  :)

**Comment by gondaruk** [comment]
Looks like before gcc was introduced, `target_bitrate` was used in calculation which was [always](https://github.com/centricular/webrtcsink/blob/64f664c859af180679bebc10ed44ee78793d54f5/plugins/src/webrtcsink/imp.rs#L1024) <= max_bitrate.
Here, looks like actual `bitrate` is used, which can be more than `max_bitrate`.

**Comment by gondaruk** [comment]
Hmm. My bad. `bitrate` in this context is estimated bitrate notified by `rtpgccbwe`. Maybe, you're right and `rtpgccbwe` should not estimate more than `max_bitrate`?

**Comment by gondaruk** [comment]
@MathieuDuponchelle sorry for confusion. I should have check deeper before doing PR.

Here is what I found.
If `do_fec` is set to true - `rtpgccbwe`'s `max-bitrate` [will be](https://github.com/centricular/webrtcsink/blob/main/plugins/src/webrtcsink/imp.rs#L1367) `max-bitrate * 1.5`.

So, `bitrate` in this callback is clamped but with different "max", because it includes the bitrate which is supposed to be used for fec.
I've updated the PR to contain `fec` check as well.

Let me know what you think. Thanks :)

**Comment by MathieuDuponchelle** [comment]
@thiblahute ? this is all a bit ugly, not sure how to make nicer :)

**Comment by thiblahute** [comment]
I think the way to make it nicer would be to have the `max-bitrate` take FEC into account all the time and then set the encoder bitrate / fec porcentage as needed, what do you think?

**Comment by MathieuDuponchelle** [comment]
> I think the way to make it nicer would be to have the `max-bitrate` take FEC into account all the time and then set the encoder bitrate / fec porcentage as needed, what do you think?

You mean that it should also be the meaning of the prop as exposed to the user? I suppose that's a solution

**Comment by MathieuDuponchelle** [comment]
@gondaruk , I guess that needs a decision, I say we go with what thibault suggests if you agree :)

**Comment by gondaruk** [comment]
Thanks!
Let me work on that and I'll get back :)

**Comment by gondaruk** [comment]
Hey @MathieuDuponchelle @thiblahute,
whenever you have a chance, could you please take a look on the update and let me know if it's what you meant or if I misunderstood you?
Thanks!

**Comment by MathieuDuponchelle** [comment]
care to review this one @thiblahute ? :)

File: plugins/src/webrtcsink/imp.rs [hunk 1]
```diff
@@ -1749,29 +1744,28 @@ impl WebRTCSink {
         let mut state = element.imp().state.lock().unwrap();
 
         if let Some(consumer) = state.consumers.get_mut(peer_id) {
+            // Start adding some FEC when the bitrate > 2Mbps as we found experimentally
```
Comment (thiblahute) [inline plugins/src/webrtcsink/imp.rs:1747]: Maybe add that comment around the constant definition?

File: plugins/src/webrtcsink/imp.rs [hunk 2]
```diff
@@ -1749,29 +1744,28 @@ impl WebRTCSink {
...
+            // that it is not worth it below that threshold
             let fec_ratio = {
-                // Start adding some FEC when the bitrate > 2Mbps as we found experimentally
-                // that it is not worth it below that threshold
-                if bitrate <= 2_000_000 || consumer.cc_info.max_bitrate <= 2_000_000 {
+                if bitrate <= DO_FEC_THRESHOLD || consumer.cc_info.max_bitrate <= DO_FEC_THRESHOLD {
```
Comment (thiblahute) [inline plugins/src/webrtcsink/imp.rs:1750]: `consumer.cc_info.max_bitrate <= DO_FEC_THRESHOLD` should be unnecessary as bitrate can't be > max_bitrate?

File: plugins/src/webrtcsink/imp.rs [hunk 3]
```diff
@@ -1749,29 +1744,28 @@ impl WebRTCSink {
...
-                    (bitrate as f64 - 2_000_000.)
-                        / (consumer.cc_info.max_bitrate as f64 - 2_000_000.)
+                    (bitrate as f64 - DO_FEC_THRESHOLD as f64)
+                        / (consumer.cc_info.max_bitrate as f64 - DO_FEC_THRESHOLD as f64)
                 }
             };
 
-            let fec_percentage = fec_ratio * 50f64;
-            let encoders_bitrate = ((bitrate as f64) / (1. + (fec_percentage / 100.))) as i32;
+            let fec_percentage = (fec_ratio * 50f64) as u32;
```
Comment (thiblahute) [inline plugins/src/webrtcsink/imp.rs:1758]: My understanding is that `bitrate` is the full bitrate for a particular stream, including the FEC percentage, in practice this means that we need to set the `encoder_bitrate` as bitrate / fec_percentage as it was the case before.'

Why do you think it is not the case anymore?

(we should add unit tests :D)

File: plugins/src/webrtcsink/imp.rs [hunk 4]
```diff
@@ -1749,29 +1744,28 @@ impl WebRTCSink {
...
 
             if let Some(ref rtpxsend) = consumer.rtprtxsend.as_ref() {
                 rtpxsend.set_property("stuffing-kbps", (bitrate as f64 / 1000.) as i32);
             }
 
             for encoder in consumer.encoders.iter_mut() {
-                encoder.set_bitrate(element, encoders_bitrate);
+                encoder.set_bitrate(element, bitrate as i32);
```
Comment (thiblahute) [inline plugins/src/webrtcsink/imp.rs:1765]: This reminds me that here we do not take into account the number of encoders. I think that  `webrtcsink::max-bitrate` should be the maximal bitrate **per stream**, meaning that when we set it to the bandwidth estimator we should do `max-bitrate = webrtcsink.max_bitrate * consumer.encoders.len()` and here we should end up with:

```
encoder_bitrate = ((((bitrate as f64) / (1. + (fec_percentage / 100.)))) / consumer.encoders.len()) as i32
```

@MathieuDuponchelle do you agree?

File: plugins/src/webrtcsink/imp.rs (hunk 1 above)
Comment (gondaruk) [reply plugins/src/webrtcsink/imp.rs:1747]: Makes sense. Will do :)

File: plugins/src/webrtcsink/imp.rs (hunk 2 above)
Comment (gondaruk) [reply plugins/src/webrtcsink/imp.rs]: Makes sense. I've also noticed that `do_fec` wasn't considered here, so part of bitrate for fec was always substracted from the encoders bitrate (even if fec is disabled).

File: plugins/src/webrtcsink/imp.rs (hunk 4 above)
Comment (gondaruk) [reply plugins/src/webrtcsink/imp.rs]: I tried to do that, however, at the moment, when `rtpgccbwe` is created, `session.encoders` vector is still empty. I'd appreciate, if you could advise how/if this should be solved.

File: plugins/src/webrtcsink/imp.rs (hunk 3 above)
Comment (gondaruk) [reply plugins/src/webrtcsink/imp.rs]: I don't :) 
You're right, fixed that. Thank you!

File: plugins/src/webrtcsink/imp.rs (hunk 4 above)
Comment (thiblahute) [reply plugins/src/webrtcsink/imp.rs]: I think we need to keep a ref to the bandwith estimating element and set its `max-bitrate` each time an encoder is added.

**Comment by gondaruk** [comment]
@thiblahute @MathieuDuponchelle, whenever you have a chance, could you please take a look?
Thanks :)

File: plugins/src/webrtcsink/imp.rs [hunk 5]
```diff
@@ -1362,48 +1376,49 @@ impl WebRTCSink {
...
-                                return None;
-                            },
-                            Ok(e) => {
-                                let max_bitrate = (settings.cc_info.max_bitrate as f64 * if settings.do_fec { 1.5 } else { 1.}) as u32;
-                                e.set_properties(&[
+                let rtpgccbwe = match gst::ElementFactory::make("rtpgccbwe", None) {
+                    Err(err) => {
+                        glib::g_warning!(
+                            "webrtcsink",
+                            "The `rtpgccbwe` element is not available \
```
Comment (thiblahute) [inline plugins/src/webrtcsink/imp.rs:1385]: Indentation looks fishy, did you run `cargo fmt`?

File: plugins/src/webrtcsink/imp.rs [hunk 6]
```diff
@@ -1362,48 +1376,49 @@ impl WebRTCSink {
...
+                                            not doing any congestion control: {err:?}"
+                        );
+                        None
+                    }
+                    Ok(cc) => {
+                        webrtcbin.connect_closure(
+                            "request-aux-sender",
+                            false,
+                            glib::closure!(@watch element, @strong session_id, @weak-allow-none cc
```
Comment (thiblahute) [inline plugins/src/webrtcsink/imp.rs:1394]:
```suggestion
                            glib::closure!(@watch element, @strong session_id, @watch cc
```

Maybe?

**Review by thiblahute — COMMENTED** [review]
This looks good to me.

File: plugins/src/webrtcsink/imp.rs (hunk 5 above)
Comment (gondaruk) [reply plugins/src/webrtcsink/imp.rs:1385]: Yes. However, looks like `cargo fmt` don't mind if I manually set second line indentation.
First argument `"webrtcsink"` though is always moved to the new line. Not sure why.

Feel free to reopen the conversation if updated version doesn't look good to you ;)

File: plugins/src/webrtcsink/imp.rs (hunk 6 above)
Comment (gondaruk) [reply plugins/src/webrtcsink/imp.rs:1394]: So am I :) 
Thanks for review!

File: plugins/src/webrtcsink/imp.rs (hunk 4 above)
Comment (gondaruk) [reply plugins/src/webrtcsink/imp.rs]: Thanks for suggestion. 
This was basically done. Feel free to reopen the conversation, if needed :)

**Comment by gondaruk** [comment]
Updated PR heading to reflect what was implemented at the end.

Let me know, if I can do anything else to get this merged. Thanks :)

**Comment by gondaruk** [comment]
@MathieuDuponchelle, kindly reminder about this PR :)

**Comment by MathieuDuponchelle** [comment]
OK let's go then :)

**Comment by MathieuDuponchelle** [comment]
Thanks again @gondaruk :)

**Pull request merged**